src/java.base/share/classes/jdk/internal/classfile/impl/ClassHierarchyImpl.java

```diff
@@ -124,9 +128,11 @@ public CachedClassHierarchyResolver(Function<ClassDesc, InputStream> classStream
         // empty ClInfo is stored in case of an exception to avoid repeated scanning failures
         @Override
         public ClassHierarchyResolver.ClassHierarchyInfo getClassInfo(ClassDesc classDesc) {
-            var res = resolvedCache.get(classDesc);
-            //additional test for null value is important to avoid repeated resolution attempts
-            if (res == null && !resolvedCache.containsKey(classDesc)) {
+            //using NOPE to distinguish between null value and non-existent record in the cache
+            //this code is on JDK bootstrap critical path, so cannot use lambdas here
+            var res = resolvedCache.getOrDefault(classDesc, NOPE);
```

Comment:
`resolvedCache` is a `Collections.synchronizedMap` backed by a `HashMap`. So the `getOrDefault()` (internally) synchronizes on the `resolvedCache` instance itself and thus this operation would be thread safe in context of access to this cache in rest of this code. So what you propose here looks good to me.

Reply:
Thank you.